### DISCUSSION TITLE
nodejs-polars in server component

### DISCUSSION THREAD

**Discussion by elge70** [vercel/next.js#63884] [Help]
### Summary

I am using nodejs-polars in my server component for data manupulation. I am using following plugin to include the binary module in next.config.js

    webpack: (config, { dev, isServer, webpack, nextRuntime }) => {
      config.module.rules.push({
        test: /\.node$/,
        use: [
          {
            loader: "nextjs-node-loader",
            options: {
              flags: os.constants.dlopen.RTLD_NOW,
              outputPath: config.output.path,
            },
          },
        ],
      });
      return config;
    },

But when I try to access the component in server, I get following error 
"Error: Cannot find module './nodejs-polars.win32-x64-msvc.node'"

I am running my development environment on Windows 11 and node 21.7.0. I tried to set NODE_PATH to node_modules where nodejs-polars is installed. Still I get this error.

Could you please help

**Comment by archnode** [comment]
I have the same issue with nodejs-polars in my project. I'm currently using the webpack-based build configuration which works with my setup - my configuration is basically identical to yours (I'm on linux, though).

One issue I couldn't work around was with symlinked packages via `npm link` that require nodejs-polars. This always led to an error similar to the one you reported.

**Reply by lamberto15** [reply]
Here's what i did to make it work. Hope it helps.

`
webpack: (config, { isServer, webpack }) => {
        if (isServer) {
            config.externals.push({
                'nodejs-polars': 'commonjs nodejs-polars'
            });
        }
        return config
}
`

**Reply by sherlockwang** [reply]
You are my hero. It stuck me more than a month. How do you find out the solution?

**Reply by lamberto15** [reply]
By using config.externals.push(), the code is telling Webpack not to bundle a certain module (nodejs-polars) into the server-side bundle. Instead, it treats nodejs-polars as an external dependency, meaning that Webpack will expect it to be available at runtime (on the server), rather than bundling it.

I've also been stuck for several days, but I'm really determined to use dataframes with Next.js. That's why I’m searching for ways to integrate them into my Next.js project. 😁😁😁

**Reply by sherlockwang** [reply]
Thanks for the really clear explanation. It not only solves the server component issue, but also remove module not found error during building, which make my project way stronger and doesn't terminate the thread randomly.
